Add geoshape query documentation (#7829)
opensearch-project/documentation-website: _query-dsl/geo-and-xy/geoshape.md

---
layout: default
title: Geoshape
parent: Geographic and xy queries
nav_order: 40
---

# Geoshape query

Use a geoshape query to search for documents that contain [geopoint]({{site.url}}{{site.baseurl}}/opensearch/supported-field-types/geo-point/) or [geoshape]({{site.url}}{{site.baseurl}}/opensearch/supported-field-types/geo-shape/) fields. You can filter documents using a [geoshape that is defined within a query](#using-a-new-shape-definition) or use a [pre-indexed geoshape](#using-a-pre-indexed-shape-definition).

The searched document field must be mapped as `geo_point` or `geo_shape`.
{: .note}

## Spatial relations

When you provide a geoshape to the geoshape query, the geopoint and geoshape fields in the documents are matched using the following spatial relations to the provided shape.

Relation | Description | Supporting geographic field type
:--- | :--- | :--- 

> `INTERSECTS` | (Default) Matches documents whose geopoint or geoshape intersects the shape provided in the query. | `geo_point`, `geo_shape`

Author: Adding "with" for consistency here. I prefer to use "intersects with" throughout.

:--- | :--- | :---
`id` | Required | The document ID of the document that contains the pre-indexed shape. 
`index` | Optional | The name of the index that contains the pre-indexed shape. Default is `shapes`.
`path` | Optional | The field name of the field that contains the pre-indexed shape as a path. Default is `shape`.
`routing` | Optional | The routing of the document containing the pre-indexed shape.

The following example illustrates referencing the name of a shape pre-indexed in another index. In this example, the index `pre-indexed-shapes` contains the shape that defines the boundaries, and the index `testindex` contains the shapes that are checked against those boundaries.

First, create the `pre-indexed-shapes` index and map the `boundaries` field for this index as a `geo_shape`:

```json
PUT /pre-indexed-shapes
{
  "mappings": {
    "properties": {
      "boundaries": {
        "type": "geo_shape",
        "orientation" : "left"
      }
    }
  }
}
```
{% include copy-curl.html %}

For more information about specifying a different vertex orientation for polygons, see [Polygon]({{site.url}}{{site.baseurl}}/field-types/supported-field-types/geo-shape/#polygon). 

> Index a polygon that specifies the search boundaries into the `pre-indexed-shapes` index. The polygon's ID is `search_triangle`. In this example, you'll index the polygon in WKT format:

Author: Reworded.

> Index a polygon that specifies the search boundaries into the `pre-indexed-shapes` index. The polygon's ID is `search_triangle`. In this example, you'll index the polygon in WKT format:

Author: ```suggestion
Index a polygon specifying the search boundaries into the `pre-indexed-shapes` index. The polygon's ID is `search_triangle`. In this example, you'll index the polygon in WKT format:
```

    "successful": 1,
    "skipped": 0,
    "failed": 0
  },
  "hits": {
    "total": {
      "value": 2,
      "relation": "eq"
    },
    "max_score": 1,
    "hits": [
      {
        "_index": "testindex",
        "_id": "1",
        "_score": 1,
        "_source": {
          "location": {
            "type": "point",
            "coordinates": [
              73.0515,
              41.5582
            ]
          }
        }
      },
      {
        "_index": "testindex",
        "_id": "2",
        "_score": 1,
        "_source": {
          "location": {
            "type": "polygon",
            "coordinates": [
              [
                [
                  73.0515,
                  41.5582
                ],
                [
                  72.6506,
                  41.5623
                ],
                [
                  72.6734,
                  41.7658
                ],
                [
                  73.0515,
                  41.5582
                ]
              ]
            ]
          }
        }
      }
    ]
  }
}
```

## Querying geopoints

You can also use a geoshape query to search for documents that contain geopoints. 

> Geoshape queries on geopoints only support the default `INTERSECTS` spatial relation, so you don't need to provide the `relation` parameter.

Author: I can change to "on geopoint fields". With or containing is not the right meaning 😄

> Geoshape queries on geopoints only support the default `INTERSECTS` spatial relation, so you don't need to provide the `relation` parameter.

Author: ```suggestion
Geoshape queries on geopoint fields only support the default `INTERSECTS` spatial relation, so you don't need to provide the `relation` parameter.
```

> Geoshape queries on geopoints only support the default `INTERSECTS` spatial relation, so you don't need to provide the `relation` parameter.

Reviewer: That's much clearer, thanks :-)

{: .note}

{: .important }

> > Geoshape queries on geopoints do not support the following geoshapes:

Author: ```suggestion
> Geoshape queries on geopoint fields do not support the following geoshapes:
```

> 
> - Point
> - Linestring

> > - Multipoint

Reviewer: ```suggestion
> - Multipoints
```

> > - Multilinestring

Reviewer: ```suggestion
> - Multilinestrings
```

> - Geometry collection that contains one of the preceding geoshape types

Create a mapping where `location` is a `geo_point`:

```json
PUT /testindex1
{
  "mappings": {
    "properties": {
      "location": {
        "type": "geo_point"
      }
    }
  }
}
```
{% include copy-curl.html %}

Index two points into the index. In this example, you'll provide the geopoint coordinates as strings:

```json
PUT /testindex1/_doc/1
{
  "location": "41.5623, 72.6506"
}
```
{% include copy-curl.html %}

```json
PUT /testindex1/_doc/2
{
  "location": "76.0254, 39.2467" 
}
```
{% include copy-curl.html %}

 For information about providing geopoint coordinates in various formats, see [Formats]({{site.url}}{{site.baseurl}}/field-types/supported-field-types/geo-point/#formats). 

> Search for geopoints that intersect with the provided polygon:

Reviewer: ```suggestion
Search for geopoints that intersect the provided polygon:
```

```json
GET /testindex1/_search
{
  "query": {
    "geo_shape": {
      "location": {
        "shape": {
          "type": "polygon",
          "coordinates": [
            [
              [74.0060, 40.7128], 
              [73.7562, 42.6526], 
              [71.0589, 42.3601], 
              [74.0060, 40.7128]
            ]
          ]
        }
      }
    }
  }
}

```
{% include copy-curl.html %}

The response returns document 1:

```json
{
  "took": 21,
  "timed_out": false,
  "_shards": {
    "total": 1,
    "successful": 1,
    "skipped": 0,
    "failed": 0
  },
  "hits": {
    "total": {
      "value": 1,
      "relation": "eq"
    },
    "max_score": 0,
    "hits": [
      {
        "_index": "testindex1",
        "_id": "1",
        "_score": 0,
        "_source": {
          "location": "41.5623, 72.6506"
        }
      }
    ]
  }
}
```

> Note that when you indexed the geopoints, you specified their coordinates in the `"latitude, longitude"` format. When you search for matching documents, the coordinate array is in the `[longitude, latitude]` format. Thus, document 1 is returned in the results but document 2 is not.

Reviewer: ```suggestion
Note that when you indexed the geopoints, you specified their coordinates in `"latitude, longitude"` format. When you search for matching documents, the coordinate array is in `[longitude, latitude]` format. Thus, document 1 is returned in the results but document 2 is not.
```

> Note that when you indexed the geopoints, you specified their coordinates in the  `"latitude, longitude"` format. When you search for matching documents, the coordinate array is in the `[longitude, latitude]` format. Thus, document 1 is returned in the results but document 2 is not.

Reviewer 2: ```suggestion
Note that when you indexed the geopoints, you specified their coordinates in the `"latitude, longitude"` format. When you search for matching documents, the coordinate array is in the `[longitude, latitude]` format. Thus, document 1 is returned in the results but document 2 is not.
```

## Request fields

Geoshape queries accept the following fields.

Field | Data type | Description
:--- | :--- | :--- 

> `ignore_unmapped` | Boolean | Specifies whether to ignore an unmapped field. If set to `true`, then the query does not return any documents that contain an unmapped field. If set to `false`, then an exception is thrown when the field is unmapped. Optional. Default is `false`.

Reviewer: ```suggestion
`ignore_unmapped` | Boolean | Specifies whether to ignore an unmapped field. If set to `true`, then the query does not return any documents that contain an unmapped field. If set to `false`, then an exception is thrown when a field is unmapped. Optional. Default is `false`.
```
